Add arrayofstruct linter

saschagrunert opened the pull request:
If you have an array of a struct type, the struct must have at least one required field. This commit adds the linter which is also capable of providing a suggested fix.

Fixes https://github.com/kubernetes-sigs/kube-api-linter/issues/28

JoelSpeed reviewed (commented):
This is a great start, please take a look at how the package is registered in the registration package and the documentation that is needed and I think we will be fairly close to getting this ready to go

Do you plan to allow any configuration of the desired `required` marker to use?

saschagrunert commented:
> This is a great start, please take a look at how the package is registered in the registration package and the documentation that is needed and I think we will be fairly close to getting this ready to go

Thanks! I updated `pkg/registration/doc.go` and updated `docs/linters.md` to ship the docs.

> Do you plan to allow any configuration of the desired `required` marker to use?

I removed the suggested fix functionality and would say we can keep it simple for now. 

The current approach accept any of the three standard required markers (+required, +kubebuilder:validation:Required, +k8s:required) and just check that at least one exists. Do we need more flexibility?

JoelSpeed commented:
> I updated pkg/registration/doc.go and updated docs/linters.md to ship the docs.

I don't see this pushed at the moment?

JoelSpeed commented:
/lgtm
/approve

Thanks for the quick turnaround on feedback @saschagrunert

everettraven commented:
Whoops - I guess  i was too slow on my review 😅

saschagrunert commented on pkg/analysis/arrayofstruct/analyzer.go:
```diff
@@ -0,0 +1,211 @@
+	case *ast.StructType:
+		// Inline struct definition
+		return et
+	case *ast.Ident:
```
Added a recursion end on basic type in https://github.com/kubernetes-sigs/kube-api-linter/pull/166

saschagrunert commented on pkg/analysis/arrayofstruct/analyzer.go:
```diff
@@ -0,0 +1,211 @@
+}
+
+// isObjectType checks if the given expression represents an object type (not a primitive).
+func isObjectType(pass *analysis.Pass, expr ast.Expr) bool {
```
I wasn't too successful in finding similar places for this function, so I decided to keep that change out of https://github.com/kubernetes-sigs/kube-api-linter/pull/166 for now.
